src/storage/meta.rs

```diff
@@ -0,0 +1,36 @@
+
+use std::marker::PhantomData;
+
+use Index;
+
+/// Observes interactions with the component's storage and stores state alongside it.
```

Comment:
I think `Metadata` is too broad for this. Perhaps, something like `Hook` would do better? Or maybe `Observer`.